RFC-123: GitHub Actions CI (#123)
alphagov/govuk-rfcs: rfc-123-github-actions-ci.md

- Already in use across some of our production applications. Examples: [business volunteer form][business-volunteer-form], [find support form][find-support-form], [vulnerable people form][vulnerable-people-form]
- In the future, we can make our internal GitHub Actions into libraries so we’re not duplicating “set up Rails apps” steps across 50 apps and can update them in one place (perhaps `alphagov/actions`?). We already do something similar with `alphagov/govuk-jenkinslib`. See [Homebrew/actions](https://github.com/Homebrew/actions) for an example.

### Cons

- We have less control over what software is on the runners. But most of the time we only need a database, a version of Ruby and some gems. And updating our Jenkins CI instances with new packages is a massive pain.
- Another system to learn, as well as the replatforming work already ongoing.
- Duplication of build parameters and steps between all our apps, as they’re all Rails apps. This can be mitigated if we use a shared actions library as proposed earlier.

### Findings from some initial explorations into using GitHub Actions

#### `push` vs `pull_request` build triggers

We considered the following options:

> 1. Running `on: [pull_request, push]`. This means that internal devs get feedback on their commits when they push them, for example if they've forgotten to run RuboCop locally. And also so that external contributors PRs get CI feedback on their changes. It does mean that for long-running test suites, we could waste compute power, time and the planet's precious natural resources running things twice that don't need to be run twice.

Reviewer: ```suggestion
1. Running `on: [pull_request, push]`. This means that internal devs get feedback on their commits when they push them, even if they haven't opened a PR yet (via `push`). External contributors can also get CI feedback on their PRs (via `pull_request`). It does mean that pushing commits to an open PR will waste compute power, time and the planet's precious natural resources, since we'll be running things twice (via `push` and `pull_request`).
```

> 2. Running `on: push`. Actions don't run [by default on forked PRs](https://github.com/alphagov/govuk-docker/pull/337) - it's a security feature that GitHub has implemented. And [when a PR is raised from a fork](https://github.com/alphagov/govuk-docker/pull/337) and the receiving repo only has a push trigger, the tests don't run. So we'd be back to the annoying state that is [our current process for forked PRs][merge-a-pull-request].

Reviewer: ```suggestion
2. Running `on: push`. This would persist a current problem we have with running tests on PRs from external contributors. When [a PR is raised from a fork](https://github.com/alphagov/govuk-docker/pull/337), the tests won't run (needs `pull_request`). Actions also [don't run on forked PRs](https://github.com/alphagov/govuk-docker/pull/337). So we'd be back to the annoying state that is [our current process for forked PRs][merge-a-pull-request].
```

> 3. Running `on: pull_request`. Functionally, we run the same tests across the board, whichever of these we choose. But this one has the advantage that it's consistent behaviour for internal and external contributors. But internally we'd have to put up with not having tests run on branches (apart from `master`, we should probably include the `master` branch all the time), which would mean changing developer workflows to run more tests locally (ie RuboCop, a common thing that gets forgotten about), or being less worried about "exposing to the world" our force-pushes to get things right on the PR. We would also probably benefit from opening PRs for long-running branches earlier, using the draft PR functionality, and this would encourage that. This seems like the best option to _Issy_, but please feel free to disagree.

Reviewer: ```suggestion
3. Running `on: pull_request`. This has the advantage that it's consistent behaviour for internal and external contributors. Internally we'd have to put up with not having tests run on branches, prior to raising a PR. This would mean changing developer workflows to run more tests locally, or being less worried about "exposing to the world" our force-pushes to get things right on the PR. (Irrespective of this decision, we should always build the `master` branch when merge commits are `push`ed.)
```

In a [poll of all interested RFC respondents](https://github.com/alphagov/govuk-rfcs/pull/123#issuecomment-620028907), option 1 - running builds on both `pull_request` and `push` events - was the most popular. In that case, going forward, all Actions CI builds will run `on: [push, pull_request]`.

#### `release` tagging via GitHub Actions

Currently, our release tags are pushed [at the end of a Jenkins `master` build](https://github.com/alphagov/govuk-jenkinslib/blob/master/vars/govuk.groovy#L705-L727). The numbers auto-increment each time. These have been reset to 0 on a few occasions during Jenkins upgrades. A consequence of tagging releases on successful master builds via GitHub Actions is that our `release` tags would be reset to 0 once again.

### The state of our repos at the moment

Currently we have:

- The vast majority of our repos run CI checks on Jenkins.
- The new apps we built recently (the coronavirus forms), and the Seal (as it's not an "app" per-se, it's deployed to Heroku), use GitHub Actions.
- Some of them (DGU) run CI on Travis.
- Some GOV.UK Data Labs projects use Concourse for PR checks.

## Outcome and next steps

> We will gradually migrate all of the GOV.UK repos to run CI builds with GitHub Actions. To start with, these could be the [`skipdeployToIntegration`][] repos, or tooling. We need to [make `govuk-saas-config` set GitHub Actions checks to "required"][govuk-saas-config].

Reviewer: ```suggestion
We will gradually migrate all of the GOV.UK repos to run CI builds with GitHub Actions. To start with, these could be the [`skipDeployToIntegration`][] repos, or tooling. We need to [make `govuk-saas-config` set GitHub Actions checks to "required"][govuk-saas-config].
```

- Concourse CI (self-hosted in TechOps, behind IP and GitHub team membership restrictions)
- GitHub Actions CI (hosted, paid through GDS' GitHub Enterprise Cloud subscription)

## Proposal

To switch to GitHub Actions CI for continuous integration (that is, branch and PR builds for tests/linting) as the preferred approach where we do not need the wider platform integration of Jenkins.

This would involve working towards GOV.UK projects only using either GitHub Actions or Jenkins for CI, no projects should use Concourse, Travis or another tool for CI without reasons to justify an exception.

This would be a gradual process, because we have many repos, but this decision would deem the usage of Travis CI as a legacy approach to CI for GOV.UK.

### Pros

- A simpler, more modern approach compared to Jenkins.
- GDS as a whole pays for GitHub Enterprise Cloud, which entitles us to unlimited minutes on public repos (both Linux and macOS runners), and 50,000 minutes per month of Actions on private repos.
- Supported by GitHub. We get Enterprise support via our GitHub Enterprise Cloud subscription. And we don't have to host any of our own infrastructure.

> - Well documented, easy to [get started][actions-get-started] building a new pipeline. Easier than Concourse for CI as PR builds are built-in to GitHub rather than needing to use third-party code]().

Reviewer 2: Nitpick - are these brackets meant to be here?

```suggestion
- Well documented, easy to [get started][actions-get-started] building a new pipeline. Easier than Concourse for CI as PR builds are built-in to GitHub rather than needing to use third-party code.
```

Author: Agh, I thought I meant to include a link but it turns out I didn't.

> - Well documented, easy to [get started][actions-get-started] building a new pipeline. Easier than Concourse for CI as PR builds are built-in to GitHub rather than needing to use third-party code]().

Author: ```suggestion
- Well documented, easy to [get started][actions-get-started] building a new pipeline. Easier than Concourse for CI as PR builds are built-in to GitHub rather than needing to use third-party code](https://github.com/telia-oss/github-pr-resource).
```

> - Well documented, easy to [get started][actions-get-started] building a new pipeline. Easier than Concourse for CI as PR builds are built-in to GitHub rather than needing to use third-party code](https://github.com/telia-oss/github-pr-resource).

Reviewer 3: I think you're missing an `[` somewhere on this line

Author: Argh! Thanks. 😓 I think I fixed it now.

- External contributors can see their changes and why they failed, without us having to [push branches][merge-a-pull-request] and paste build logs from Jenkins. This fits more readily with our "code in the open" ethos, and saves on internal developer time.
- Already in use across [some][business-volunteer-form] of [our][find-support-form] [production][vulnerable-people-form] applications.
- As we go further down the path, we can make our internal GitHub Actions into libraries so we’re not duplicating “set up Rails apps” steps across 50 apps and can update them in one place - like we do with `alphagov/govuk-jenkinslib` now. Maybe `actions/alphagov`? (See [Homebrew/actions](https://github.com/Homebrew/actions) for an example.)
- Getting rid of Jenkins would be a huge cost saving - our CI environment is in Carrenza and currently costs ~£6k a month + VAT.

### Cons

- We have less control over what software is on the runners. But most of the time we only need a database, a version of Ruby and some gems. And updating our Jenkins CI instances with new packages is a massive pain.

> - How do we do credentials management for things like pushing draft releases and Docker images to DockerHub?

Reviewer 3: I think any CI process that uses secrets will have to run on Concourse to get all of the assurances we need.

Author: We should. @bevanloon @richardTowers Can we have a chat about this? It also relates to [this gem publishing comment](https://github.com/alphagov/rubocop-govuk/pull/41#issuecomment-618378751). (We're already doing gem publishing with GitHub Actions, and it's working well.) There's some detailed discussion on GitHub Actions secrets that I linked to in the above `rubocop-govuk` PR - that's worth a read.

Author: That said, can anyone think of any CI (ie linting and app tests - as opposed to *deployment processes*) that need secrets to run?

Reviewer: I'm not aware of anything. However, if we don't allow secrets, do we have an extra consequence that gem releases need to happen <somewhere else>?

Reviewer 4: I think it would be easier to make the "no secrets" decision for now - that still lets us do the majority of CI things, but as Ben says, we'll need to decide where to do gem releases (because we can't really do those in GitHub Actions without doing the secrets / IA / cyber dance).

Reviewer: 👍 - sounds like we'll be wanting to revert part of the demo for [rubocop-govuk](https://github.com/alphagov/rubocop-govuk/pull/45) and go with [some kind of restricted Jenkins build](https://github.com/alphagov/govuk-jenkinslib/pull/70) instead. That does seem like a slightly worse situation than what we had before, but I guess it only applies to a few repos - we only have a few gems.

Reviewer 5: The pact contract testing process requires secrets to publish new pacts (used by Publishing API, GDS API Adapters and maybe Content Store).

I do think we'd be creating quite a pain point if we go down a no secrets route and would mean that we retrospectively need to undo some travis jobs we have that release npm modules. So I do think we should resolve it rather than going down "no secrets" as a simple option.

Author: Can we go forwards not backwards, please? I'd also argue that any gem publishing we do on Jenkins (if we must) should be done on the deploy Jenkinses rather than the CI Jenkinses - for separation of concerns.

Author: (Also my apologies, that previous comment came across a bit angry.)

Reviewer 4: Maybe "no secrets" is the wrong heuristic. I think the gem ones are sensitive because there's a possible vector for compromising GOV.UK in there:

1) steal the rubygems secret from GitHub actions somehow
2) publish a malicious version of a gem that we use
3) wait for dependabot to update it
4) ?????
5) very bad things

Not to say that won't be an acceptable risk once we've spoken to cyber and IA (depends on how we feel about GitHub actions' security), but they definitely need to be in the loop.

The pact example is much less problematic, as it's not going to affect production code.

Also worth noting that I don't actually understand how we used to publish our gems, so I'm not sure what going forwards and backwards actually means...

Reviewer 5: > Maybe "no secrets" is the wrong heuristic. I think the gem ones are sensitive because there's a possible vector for compromising GOV.UK in there:
> 
> 1. steal the rubygems secret from GitHub actions somehow
> 2. publish a malicious version of a gem that we use
> 3. wait for dependabot to update it
> 4. ?????
> 5. very bad things
> 
> Not to say that won't be an acceptable risk once we've spoken to cyber and IA (depends on how we feel about GitHub actions' security), but they definitely need to be in the loop.
> 
> The pact example is much less problematic, as it's not going to affect production code.
> 
> Also worth noting that I don't actually understand how we used to publish our gems, so I'm not sure what going forwards and backwards actually means...

I'd suggest that same vector applies for any gem and even our ones currently as there are already means that either our account or RubyGems is compromised. So that vector should be covered by always verifying what you're merging with dependabot (whether people always do is another matter)

I'd describe our risk as reputational, since were a token to be compromised then GOV.UK's name would be associated with it.

On something of a tangent the move to GitHub actions might also reduce security risks from a different angle as, for apps enabled, we would be building dependabot PRs on GitHubs infrastructure rather than our own.

Author: OK so I had a chat with Richard offline (thanks!) and we would like to reiterate that this RFC was only supposed to be about CI (branch builds for testing and linting). For *this case* GitHub Actions is fine. For gems and projects that aren't production code, GitHub Actions is fine for releasing too. For instance, `rubocop-govuk` is only a dev and test gem - and Kevin's `npm` pushes are fine as `html-to-govspeak` isn't a production service either, so it's fine to keep them as-is.

There may well be extensions to this RFC in future when we're at the point where IA/Cyber have done their assessments into GitHub Actions secrets security for *production* secrets, but that's the future. The *now* on this is GitHub Actions for CI builds, to save us a tonne of money and increase developer productivity.

Thank you. Now where's the "also send to channel" button on for GitHub threads...

- Another system to learn, as well as the replatforming work already ongoing.
- Duplication of build parameters and steps between all our apps, as they’re all Rails apps.

### The state of our repos at the moment

Currently we have:

- The vast majority of our repos run CI checks on Jenkins.
- The new apps we built recently (the coronavirus forms), and the Seal (as it's not an "app" per-se, it's deployed to Heroku), use GitHub Actions.
- Some of them (DGU) run CI on Travis.

> - I can't find any uses of Concourse on GOV.UK repos.

Reviewer 3: A few Data Labs projects use Concourse for regular jobs, not CI

Reviewer 6: We're increasingly using Concourse for running PR tests, linting etc before they're able to be merged (nothing is run on merge itself).

Author: Ah, I missed you out! My assessments here were done based on the contents of my local `~/govuk` folder, and I don't think I have any of the data labs stuff.

# RFC-123: Use GitHub Actions for CI

## Summary

This proposes using [GitHub Actions](https://github.com/features/actions)  as our preferred CI tool where applicable.

CI in this case is defined as "running builds for tests and/or linting, on branches and PRs". This does not cover *deployment* of the master branch, which will continue to happen using Jenkins (until such time as we replatform (see RFC 117)) and will be considered as part of wider built pipeline work (such as continuous deployment).

TechOps [started a discovery into CI tooling](https://github.com/alphagov/reliability-engineering/pull/106) a while ago, but it's not reached a conclusion yet. Now we've procured GitHub Enterprise Cloud, making use of the tools available there reduces platform complexity.

## Problem

We currently have many solutions for running branch builds in GOV.UK, including:

- Jenkins CI (self-hosted, behind IP and GitHub team membership restrictions)

> - Travis CI (hosted, paid on a GPC card)

Author: A Government Procurement Card. A credit card (there's only a few in existence within GDS), it allows government to buy things that don't come invoiced.

- Well documented, easy to [get started][actions-get-started] building a new pipeline. Easier than Concourse for CI as PR builds are built-in to GitHub rather than needing to use [third-party code](https://github.com/telia-oss/github-pr-resource).
- Doesn't require access to the VPN like our Jenkins does, or membership of any GitHub teams, in order to see logs for public repos.
- External contributors can see their changes and why they failed, without us having to [push branches][merge-a-pull-request] and paste build logs from Jenkins. This fits more readily with our "code in the open" ethos, and saves on internal developer time.
- Already in use across [some][business-volunteer-form] of [our][find-support-form] [production][vulnerable-people-form] applications.
- As we go further down the path, we can make our internal GitHub Actions into libraries so we’re not duplicating “set up Rails apps” steps across 50 apps and can update them in one place - like we do with `alphagov/govuk-jenkinslib` now. Maybe `actions/alphagov`? (See [Homebrew/actions](https://github.com/Homebrew/actions) for an example.)

> - Getting rid of Jenkins would be a huge cost saving - our CI environment is in Carrenza and currently costs ~£6k a month + VAT.

Author: We have deploy Jenkinses in all six environments:

- [Integration Carrenza](https://ci-deploy.integration.publishing.service.gov.uk) (that's where we deploy Terraform from for the rest of the environments, and the CI environment)
- [Integration AWS](https://deploy.integration.publishing.service.gov.uk)
- [Staging Carrenza](https://deploy.staging.publishing.service.gov.uk)
- [Staging AWS](https://deploy.blue.staging.govuk.digital)
- Production Carrenza](https://deploy.publishing.service.gov.uk)
- Production AWS](https://deploy.blue.production.govuk.digital)

- Another system to learn, as well as the replatforming work already ongoing.
- Duplication of build parameters and steps between all our apps, as they’re all Rails apps.

### Findings from some initial explorations into using GitHub Actions

#### `push` vs `pull_request` build triggers

There has been some [chatter](https://github.com/alphagov/govuk-developer-docs/pull/2497#discussion_r415572147) about if we want `push` or `pull_request` triggers for tests to run, or both. The GOV.UK Developer Tools group has done some investigation into the differences between this, as there was some concern about if we wanted long-running builds (in the case of apps with large test suites, like whitehall) to run twice when not needed. There is not currently any way to differentiate and say "run on PRs from forks" and "run on push events to main repo branches". We'd have really liked a `pull_request_only_from_forks` setting, but that doesn't exist. It's worth noting that this isn't a problem _only_ with GitHub Actions: [Travis](https://github.com/travis-ci/travis-ci/issues/1147) has had the same pain points for years, and we currently do double builds for the apps we use Travis for.

> Some solutions to this include:

Reviewer 7: ```suggestion
Our options are:
```

### The state of our repos at the moment

Currently we have:

- The vast majority of our repos run CI checks on Jenkins.
- The new apps we built recently (the coronavirus forms), and the Seal (as it's not an "app" per-se, it's deployed to Heroku), use GitHub Actions.
- Some of them (DGU) run CI on Travis.
- I can't find any uses of Concourse on GOV.UK repos.

## Outcome and next steps

> The outcome of this is defined in the proposal. In the first instance, I imagine the work to do this will be taken on by the GOV.UK Developer Tools team - please, get involved!

Reviewer 7: I think we can remove this line.

Author: ```suggestion

```

- Duplication of build parameters and steps between all our apps, as they’re all Rails apps.

### Findings from some initial explorations into using GitHub Actions

#### `push` vs `pull_request` build triggers

There has been some [chatter](https://github.com/alphagov/govuk-developer-docs/pull/2497#discussion_r415572147) about if we want `push` or `pull_request` triggers for tests to run, or both. The GOV.UK Developer Tools group has done some investigation into the differences between this, as there was some concern about if we wanted long-running builds (in the case of apps with large test suites, like whitehall) to run twice when not needed. There is not currently any way to differentiate and say "run on PRs from forks" and "run on push events to main repo branches". We'd have really liked a `pull_request_only_from_forks` setting, but that doesn't exist. It's worth noting that this isn't a problem _only_ with GitHub Actions: [Travis](https://github.com/travis-ci/travis-ci/issues/1147) has had the same pain points for years, and we currently do double builds for the apps we use Travis for.

Some solutions to this include:

> - Running `on: [pull_request, push]`. This means that internal devs get feedback on their commits when they push them, for example if they've forgotten to run RuboCop locally. And also so that external contributors PRs get CI feedback on their changes. It does mean that for long-running test suites, we could waste compute power, time and the planet's precious natural resources running things twice that don't need to be run twice.

Reviewer 5: I've come around to be in favour of this option since they're slightly different changes. Does seem a pain that we might often have very similar tests run - but at least they'd be in parallel. 

I feel like this would become much more of a pain point though if action queues get long and we find we're waiting for one and not the other.

Author: On the Enterprise plan, we get [180 concurrent builds](https://help.github.com/en/actions/getting-started-with-github-actions/about-github-actions#usage-limits) across `alphagov`, assuming they're all Linux builds and not macOS.

Author: At the end of the day, I don't mind which option we go for (as long as it's one of "only PRs" or "both PRs and pushes", the "on push" option is in no way an improvement to our current processes), but I'd like to solicit some more opinions from others otherwise this feels a bit circular. :-)

Reviewer 8: I can see value in having test run on your branch before a PR is raised, therefore you can immediately notice a failure, rather than waiting after you've raised the PR. I feel like going with both lets us see if there's any value/downsides and we can always tweak these settings in the future.

Reviewer 7: It's worth describing that advantage in the RFC 👍

Reviewer: Have we considered another potential disadvantage: if we're running the tests twice, and they (both) fail, which one do I look at to see the failure? While this may be obvious for us, the architects of the future, I can see it being a source of confusion: which "Details" link should I click on when I see a failure?

Author: If the failures are different then we have more interesting problems. I'd look at the `push` ones, personally, as I'll be able to see quicker that it's failed. Unless it's a forked PR.

> - Running `on: push`. Actions don't run [by default on forked PRs](https://github.com/alphagov/govuk-docker/pull/337) - it's a security feature that GitHub has implemented. And [when a PR is raised from a fork](https://github.com/alphagov/govuk-docker/pull/337) and the receiving repo only has a push trigger, the tests don't run. So we'd be back to the annoying state that is [our current process for forked PRs][merge-a-pull-request].

Reviewer 5: Should be flagged here that this doesn't merge with master so we only know the branch passes and not if the integrated branch passes.

Author: GitHub does have [an option in the branch protection settings](https://help.github.com/en/github/administering-a-repository/types-of-required-status-checks) to enforce that branches are up to date with `master` before merging. But [we don't currently enforce it](https://github.com/alphagov/govuk-saas-config/blob/master/github/lib/configure_repo.rb#L95).

Reviewer 9: I don't think many people rebase before merging, so switching that on would be a big disruptive change for a lot of people.

Reviewer 7: I don't think it would be particularly disruptive actually. People on long-running branches regularly rebase against master, in my experience - and for people doing small tweaks, GitHub offers a "Rebase to master" button right on the PR itself.

The only disruption is waiting for the tests to pass again, especially if there's a flaky publishing-e2e-tests downstream too. I guess it might encourage us to make our test suites faster and more reliable.

Reviewer 6: I'm definitely an advocate of rebasing before pushing, just to make sure that everything runs fine against the latest master code. It can be disruptive in a fast-moving repo though as Chris points out above if in the time tests take to run, more commits are pushed.

> - Running `on: pull_request`. Functionally, we run the same tests across the board, whichever of these we choose. But this one has the advantage that it's consistent behaviour for internal and external contributors. But internally we'd have to put up with not having tests run on branches (apart from `master`, we should probably include the `master` branch all the time), which would mean changing developer workflows to run more tests locally (ie RuboCop, a common thing that gets forgotten about), or being less worried about "exposing to the world" our force-pushes to get things right on the PR. We would also probably benefit from opening PRs for long-running branches earlier, using the draft PR functionality, and this would encourage that. This seems like the best option to _Issy_, but please feel free to disagree.

Reviewer 5: This seems to be closest to how Jenkins runs now by always merging master into what's requested. 

It does seem annoying to lose CI though prior to opening a PR and does mean keeping the confusion of branch passing locally but failing on the integrated.

Reviewer: This is still my preferred option. I think it's confusing to see two sets of tests being run, and people running tests locally just seems easier and more efficient.

Conclusion: is the downside, really a downside, or just a habit we're trying to accommodate?

Author: I was originally in favour of the first option, running on PRs and Push events, but upon writing all the findings up, I came around to this option: only on PRs. We shouldn't be scared of showing off our work, force-pushing, or opening draft PRs - and PRs can always be closed and re-opened!

Reviewer 5: > We shouldn't be scared of showing off our work, force-pushing, or opening draft PRs - and PRs can always be closed and re-opened!

I don't think it's a factor of being scared to show work but more a question of time and efficiency. I can do some work that I expect is fine, push the branch and move onto more work and check that build passed CI. Whereas if I pause working run a full rake task that's going to add more time.

Similarly once I'm putting together a PR it's nice to know the build has passed so that once I click open it's ready for review. It's also nice that all the build process can run while I'm writing the PR.

I'm sure I'd get used to it only being on PR's but it feels like, at least for me, it'll add inefficiencies and extra time to code process. Which is something I'd rather be absorbed by CI machines than myself.

> 

Reviewer 7: Another alternative is wrapping the job in a conditional: https://github.community/t5/GitHub-Actions/don-t-run-actions-on-pull-request-from-fork/td-p/45499

```
jobs:
  prcheck:
    if: github.actor == 'repo-owner'
    runs-on: [ubuntu-latest]
    # etc
```
